### ISSUE TITLE
0.8.0: Intermittent rendering issue in GTK apps 

### ISSUE THREAD

**Issue by flexibeast** [WayfireWM/wayfire#2063]
Running 0.8.0, i've noticed a couple of odd rendering issues in GTK apps; i'm wondering if they might actually be related.

Firstly, text in some parts of an app's chrome is sometimes oddly light, resulting in low contrast:

![screencap-1](https://github.com/WayfireWM/wayfire/assets/190652/e464cc7e-4340-4934-8fe9-f87c7a82fb77)

Certain user actions result in the text being re-rendered with the appropriate value levels:

![screencap-2](https://github.com/WayfireWM/wayfire/assets/190652/c960f301-13f4-46c8-af0e-7f3288433b0a)

Some of the user actions that can do this include focusing a drop-down widget and then un-focusing it by clicking elsewhere, and - as i've just discovered - taking a screencap.

If it might help, here's the output of `ldd /usr/lib64/libreoffice/program/soffice.bin`:

```             
linux-vdso.so.1 (0x00007ffd6bb83000)
libuno_sal.so.3 => /usr/lib64/libreoffice/program/libuno_sal.so.3 (0x00007feabcd63000)
libmergedlo.so => /usr/lib64/libreoffice/program/libmergedlo.so (0x00007feab8400000)
libc.so.6 => /lib64/libc.so.6 (0x00007feab8227000)
libz.so.1 => /lib64/libz.so.1 (0x00007feab820d000)
libstdc++.so.6 => /usr/lib/gcc/x86_64-pc-linux-gnu/13/libstdc++.so.6 (0x00007feab7e00000)
libm.so.6 => /lib64/libm.so.6 (0x00007feab8130000)
libgcc_s.so.1 => /usr/lib/gcc/x86_64-pc-linux-gnu/13/libgcc_s.so.1 (0x00007feab810b000)
/lib64/ld-linux-x86-64.so.2 (0x00007feabcde0000)
libepoxy.so.0 => /usr/lib64/libepoxy.so.0 (0x00007feab7ccf000)
libgpgmepp.so.6 => /usr/lib64/libgpgmepp.so.6 (0x00007feab809a000)
libicuuc.so.73 => /usr/lib64/libicuuc.so.73 (0x00007feab7a00000)
libsmime3.so => /usr/lib64/libsmime3.so (0x00007feab8070000)
libnss3.so => /usr/lib64/libnss3.so (0x00007feab78c4000)
libplc4.so => /usr/lib64/libplc4.so (0x00007feab8069000)
libnspr4.so => /usr/lib64/libnspr4.so (0x00007feab7c8b000)
libdconf.so.1 => /usr/lib64/libdconf.so.1 (0x00007feab7c77000)
libgio-2.0.so.0 => /usr/lib64/libgio-2.0.so.0 (0x00007feab76db000)
libgobject-2.0.so.0 => /usr/lib64/libgobject-2.0.so.0 (0x00007feab7c15000)
libglib-2.0.so.0 => /usr/lib64/libglib-2.0.so.0 (0x00007feab7594000)
libdbus-1.so.3 => /usr/lib64/libdbus-1.so.3 (0x00007feab7546000)
libicui18n.so.73 => /usr/lib64/libicui18n.so.73 (0x00007feab7200000)
libcurl.so.4 => /usr/lib64/libcurl.so.4 (0x00007feab7159000)
libxml2.so.2 => /usr/lib64/libxml2.so.2 (0x00007feab6ff0000)
libxslt.so.1 => /usr/lib64/libxslt.so.1 (0x00007feab6fb0000)
libexslt.so.0 => /usr/lib64/libexslt.so.0 (0x00007feab6f98000)
libexpat.so.1 => /usr/lib64/libexpat.so.1 (0x00007feab6f6b000)
libclucene-core.so.1 => /usr/lib64/libclucene-core.so.1 (0x00007feab6db2000)
libclucene-contribs-lib.so.1 => /usr/lib64/libclucene-contribs-lib.so.1 (0x00007feab6d4c000)
libexttextcat-2.0.so.0 => /usr/lib64/libexttextcat-2.0.so.0 (0x00007feab7c0e000)
libhunspell-1.7.so.0 => /usr/lib64/libhunspell-1.7.so.0 (0x00007feab6c8c000)
libhyphen.so.0 => /usr/lib64/libhyphen.so.0 (0x00007feab753f000)
libmythes-1.2.so.0 => /usr/lib64/libmythes-1.2.so.0 (0x00007feab6c86000)
libnumbertext-1.0.so.0 => /usr/lib64/libnumbertext-1.0.so.0 (0x00007feab6c2f000)
libboost_locale.so.1.82.0 => /usr/lib64/libboost_locale.so.1.82.0 (0x00007feab6b5c000)
librdf.so.0 => /usr/lib64/librdf.so.0 (0x00007feab6b1f000)
libraptor2.so.0 => /usr/lib64/libraptor2.so.0 (0x00007feab6ab7000)
libgraphite2.so.3 => /usr/lib64/libgraphite2.so.3 (0x00007feab6a91000)
libharfbuzz-icu.so.0 => /usr/lib64/libharfbuzz-icu.so.0 (0x00007feab6a8c000)
libharfbuzz.so.0 => /usr/lib64/libharfbuzz.so.0 (0x00007feab695d000)
liblcms2.so.2 => /usr/lib64/liblcms2.so.2 (0x00007feab68f8000)
libjpeg.so.62 => /usr/lib64/libjpeg.so.62 (0x00007feab6822000)
libpng16.so.16 => /usr/lib64/libpng16.so.16 (0x00007feab67e8000)
libtiff.so.6 => /usr/lib64/libtiff.so.6 (0x00007feab6760000)
libwebp.so.7 => /usr/lib64/libwebp.so.7 (0x00007feab66f1000)
libX11.so.6 => /usr/lib64/libX11.so.6 (0x00007feab65ac000)
libXext.so.6 => /usr/lib64/libXext.so.6 (0x00007feab6597000)
libcairo.so.2 => /usr/lib64/libcairo.so.2 (0x00007feab6453000)
libcups.so.2 => /usr/lib64/libcups.so.2 (0x00007feab63b3000)
libfontconfig.so.1 => /usr/lib64/libfontconfig.so.1 (0x00007feab6366000)
libfreetype.so.6 => /usr/lib64/libfreetype.so.6 (0x00007feab629a000)
libuno_cppu.so.3 => /usr/lib64/libreoffice/program/libuno_cppu.so.3 (0x00007feab6266000)
libuno_cppuhelpergcc3.so.3 => /usr/lib64/libreoffice/program/libuno_cppuhelpergcc3.so.3 (0x00007feab6176000)
libi18nlangtag.so => /usr/lib64/libreoffice/program/libi18nlangtag.so (0x00007feab6150000)
libuno_salhelpergcc3.so.3 => /usr/lib64/libreoffice/program/libuno_salhelpergcc3.so.3 (0x00007feab6147000)
libxmlreaderlo.so => /usr/lib64/libreoffice/program/libxmlreaderlo.so (0x00007feab6139000)
libclewlo.so => /usr/lib64/libreoffice/program/libclewlo.so (0x00007feab6132000)
libskialo.so => /usr/lib64/libreoffice/program/libskialo.so (0x00007feab5a00000)
libgpgme.so.11 => /usr/lib64/libgpgme.so.11 (0x00007feab60da000)
libicudata.so.73 => /usr/lib64/libicudata.so.73 (0x00007feab3a00000)
libnssutil3.so => /usr/lib64/libnssutil3.so (0x00007feab60a5000)
libplds4.so => /usr/lib64/libplds4.so (0x00007feab59fa000)
libgmodule-2.0.so.0 => /usr/lib64/libgmodule-2.0.so.0 (0x00007feab59f3000)
libmount.so.1 => /lib64/libmount.so.1 (0x00007feab5990000)
libffi.so.8 => /usr/lib64/libffi.so.8 (0x00007feab5983000)
libpcre2-8.so.0 => /lib64/libpcre2-8.so.0 (0x00007feab58e5000)
libcares.so.2 => /usr/lib64/libcares.so.2 (0x00007feab58c6000)
libnghttp2.so.14 => /usr/lib64/libnghttp2.so.14 (0x00007feab5895000)
libssl.so.3 => /usr/lib64/libssl.so.3 (0x00007feab3958000)
libcrypto.so.3 => /usr/lib64/libcrypto.so.3 (0x00007feab3400000)
libgcrypt.so.20 => /usr/lib64/libgcrypt.so.20 (0x00007feab32b5000)
libclucene-shared.so.1 => /usr/lib64/libclucene-shared.so.1 (0x00007feab3935000)
libboost_thread.so.1.82.0 => /usr/lib64/libboost_thread.so.1.82.0 (0x00007feab3912000)
libboost_chrono.so.1.82.0 => /usr/lib64/libboost_chrono.so.1.82.0 (0x00007feab3907000)
librasqal.so.3 => //usr/lib64/librasqal.so.3 (0x00007feab3887000)
libltdl.so.7 => //usr/lib64/libltdl.so.7 (0x00007feab387b000)
libsharpyuv.so.0 => /usr/lib64/libsharpyuv.so.0 (0x00007feab3873000)
libxcb.so.1 => /usr/lib64/libxcb.so.1 (0x00007feab3848000)
libXrender.so.1 => /usr/lib64/libXrender.so.1 (0x00007feab32a8000)
libxcb-render.so.0 => /usr/lib64/libxcb-render.so.0 (0x00007feab3299000)
libxcb-shm.so.0 => /usr/lib64/libxcb-shm.so.0 (0x00007feab3843000)
libpixman-1.so.0 => /usr/lib64/libpixman-1.so.0 (0x00007feab31f0000)
libavahi-common.so.3 => /usr/lib64/libavahi-common.so.3 (0x00007feab31e2000)
libavahi-client.so.3 => /usr/lib64/libavahi-client.so.3 (0x00007feab31cf000)
libgnutls.so.30 => /usr/lib64/libgnutls.so.30 (0x00007feab2e00000)
libbz2.so.1 => /lib64/libbz2.so.1 (0x00007feab31ba000)
libreglo.so => /usr/lib64/libreoffice/program/libreglo.so (0x00007feab31a2000)
libunoidllo.so => /usr/lib64/libreoffice/program/libunoidllo.so (0x00007feab313d000)
liblangtag.so.1 => /usr/lib64/liblangtag.so.1 (0x00007feab3116000)
libX11-xcb.so.1 => /usr/lib64/libX11-xcb.so.1 (0x00007feab3111000)
libassuan.so.0 => /usr/lib64/libassuan.so.0 (0x00007feab30fb000)
libgpg-error.so.0 => /usr/lib64/libgpg-error.so.0 (0x00007feab30d1000)
libblkid.so.1 => /lib64/libblkid.so.1 (0x00007feab3078000)
libpcre.so.1 => /lib64/libpcre.so.1 (0x00007feab2d85000)
libmhash.so.2 => /usr/lib64/libmhash.so.2 (0x00007feab3048000)
libuuid.so.1 => /lib64/libuuid.so.1 (0x00007feab303c000)
libmpfr.so.6 => /usr/lib64/libmpfr.so.6 (0x00007feab2ccc000)
libXau.so.6 => /usr/lib64/libXau.so.6 (0x00007feab3036000)
libXdmcp.so.6 => /usr/lib64/libXdmcp.so.6 (0x00007feab302e000)
libidn2.so.0 => /usr/lib64/libidn2.so.0 (0x00007feab2c99000)
libunistring.so.5 => /usr/lib64/libunistring.so.5 (0x00007feab2ae8000)
libtasn1.so.6 => /usr/lib64/libtasn1.so.6 (0x00007feab3015000)
libnettle.so.8 => /usr/lib64/libnettle.so.8 (0x00007feab2a97000)
libhogweed.so.6 => /usr/lib64/libhogweed.so.6 (0x00007feab2a4b000)
libgmp.so.10 => /usr/lib64/libgmp.so.10 (0x00007feab29a9000)
libstorelo.so => /usr/lib64/libreoffice/program/libstorelo.so (0x00007feab298d000)
```

Then, running gnome-tetravex in particular, odd artifacts flicker in and out of the borders of the grid of tiles on the right-hand-side:

![screencap-3](https://github.com/WayfireWM/wayfire/assets/190652/abdfa084-e96f-4e8e-a02c-99ebf7a0c9e3)

In the above screencap, some example artifacts can be seen above the two tiles on the left of the bottom row.

Output of `ldd /usr/bin/gnome-tetravex`:

```
linux-vdso.so.1 (0x00007ffd2737f000)
libgio-2.0.so.0 => /usr/lib64/libgio-2.0.so.0 (0x00007fa434caa000)
libgobject-2.0.so.0 => /usr/lib64/libgobject-2.0.so.0 (0x00007fa434c48000)
libglib-2.0.so.0 => /usr/lib64/libglib-2.0.so.0 (0x00007fa434b01000)
libgtk-3.so.0 => /usr/lib64/libgtk-3.so.0 (0x00007fa434200000)
libgdk-3.so.0 => /usr/lib64/libgdk-3.so.0 (0x00007fa434a0e000)
libcairo.so.2 => /usr/lib64/libcairo.so.2 (0x00007fa4340bc000)
libm.so.6 => /lib64/libm.so.6 (0x00007fa433fdf000)
libc.so.6 => /lib64/libc.so.6 (0x00007fa433e06000)
libgmodule-2.0.so.0 => /usr/lib64/libgmodule-2.0.so.0 (0x00007fa434a05000)
libz.so.1 => /lib64/libz.so.1 (0x00007fa4349eb000)
libmount.so.1 => /lib64/libmount.so.1 (0x00007fa433da3000)
libffi.so.8 => /usr/lib64/libffi.so.8 (0x00007fa433d96000)
libpcre2-8.so.0 => /lib64/libpcre2-8.so.0 (0x00007fa433cf6000)
libpangocairo-1.0.so.0 => /usr/lib64/libpangocairo-1.0.so.0 (0x00007fa433ce4000)
libpango-1.0.so.0 => /usr/lib64/libpango-1.0.so.0 (0x00007fa433c78000)
libharfbuzz.so.0 => /usr/lib64/libharfbuzz.so.0 (0x00007fa433b49000)
libpangoft2-1.0.so.0 => /usr/lib64/libpangoft2-1.0.so.0 (0x00007fa433b2f000)
libfontconfig.so.1 => /usr/lib64/libfontconfig.so.1 (0x00007fa433ae2000)
libfribidi.so.0 => /usr/lib64/libfribidi.so.0 (0x00007fa433ac0000)
libcairo-gobject.so.2 => /usr/lib64/libcairo-gobject.so.2 (0x00007fa433ab5000)
libgdk_pixbuf-2.0.so.0 => /usr/lib64/libgdk_pixbuf-2.0.so.0 (0x00007fa433a86000)
libatk-1.0.so.0 => /usr/lib64/libatk-1.0.so.0 (0x00007fa433a5c000)
libepoxy.so.0 => /usr/lib64/libepoxy.so.0 (0x00007fa43392b000)
libXi.so.6 => /usr/lib64/libXi.so.6 (0x00007fa433918000)
libX11.so.6 => /usr/lib64/libX11.so.6 (0x00007fa4337d1000)
libwayland-client.so.0 => /usr/lib64/libwayland-client.so.0 (0x00007fa4337be000)
libXfixes.so.3 => /usr/lib64/libXfixes.so.3 (0x00007fa4337b6000)
libxkbcommon.so.0 => /usr/lib64/libxkbcommon.so.0 (0x00007fa43376e000)
libwayland-cursor.so.0 => /usr/lib64/libwayland-cursor.so.0 (0x00007fa433764000)
libwayland-egl.so.1 => /usr/lib64/libwayland-egl.so.1 (0x00007fa43375e000)
libXext.so.6 => /usr/lib64/libXext.so.6 (0x00007fa433749000)
libXcursor.so.1 => /usr/lib64/libXcursor.so.1 (0x00007fa43373c000)
libXdamage.so.1 => /usr/lib64/libXdamage.so.1 (0x00007fa433737000)
libXcomposite.so.1 => /usr/lib64/libXcomposite.so.1 (0x00007fa433732000)
libXrandr.so.2 => /usr/lib64/libXrandr.so.2 (0x00007fa433725000)
libpng16.so.16 => /usr/lib64/libpng16.so.16 (0x00007fa4336e9000)
libfreetype.so.6 => /usr/lib64/libfreetype.so.6 (0x00007fa43361d000)
libXrender.so.1 => /usr/lib64/libXrender.so.1 (0x00007fa433610000)
libxcb.so.1 => /usr/lib64/libxcb.so.1 (0x00007fa4335e5000)
libxcb-render.so.0 => /usr/lib64/libxcb-render.so.0 (0x00007fa4335d6000)
libxcb-shm.so.0 => /usr/lib64/libxcb-shm.so.0 (0x00007fa4335cf000)
libpixman-1.so.0 => /usr/lib64/libpixman-1.so.0 (0x00007fa433526000)
/lib64/ld-linux-x86-64.so.2 (0x00007fa434ef9000)
libblkid.so.1 => /lib64/libblkid.so.1 (0x00007fa4334cd000)
libgraphite2.so.3 => /usr/lib64/libgraphite2.so.3 (0x00007fa4334a7000)
libexpat.so.1 => /usr/lib64/libexpat.so.1 (0x00007fa43347a000)
libjpeg.so.62 => /usr/lib64/libjpeg.so.62 (0x00007fa4333a2000)
libbz2.so.1 => /lib64/libbz2.so.1 (0x00007fa43338d000)
libXau.so.6 => /usr/lib64/libXau.so.6 (0x00007fa433387000)
libXdmcp.so.6 => /usr/lib64/libXdmcp.so.6 (0x00007fa43337f000)
```

**Comment by ammen99** [comment]
The first issue was already fixed in the master branch (can you try it?), the other sounds like a gtk bug.

**Comment by ammen99** [comment]
I am unsure about the second issue, you attached just a screenshot, but I can't see anything wrong with it :)

A video of the flickering would be helpful so that I can see what is going on exactly. I just tried tetravex here locally and see no flickering though.

**Comment by flexibeast** [comment]
>  I can't see anything wrong with it :)

Fair enough. :-) There are two darker grey strips within the overall lighter grey strip above the two left tiles on the bottom row, but it's a bit of a strain to see ....

> A video of the flickering would be helpful

Indeed, and it was trying to find some way of capturing video for this issue that led my down the xdg-desktop-portal rabbithole i described in [this comment](https://github.com/WayfireWM/wayfire/issues/1995#issuecomment-1851435278) on #1995. i hardly ever need to capture video, so i'm not familiar with the relevant (Wayland) apps - can you suggest any?

> I just tried tetravex here locally and see no flickering though.

\*nod\* Thanks for checking that - i can imagine it being something related to my overall setup, or to my GTK config in particular. (Another rabbithole i've gone down in recent times is trying to work out how to configure the appearance of GTK{3,4} applications when one isn't actually running GNOME.)

**Comment by ammen99** [comment]
> Indeed, and it was trying to find some way of capturing video for this issue that led my down the xdg-desktop-portal rabbithole i described in [this comment](https://github.com/WayfireWM/wayfire/issues/1995#issuecomment-1851435278) on #1995. i hardly ever need to capture video, so i'm not familiar with the relevant (Wayland) apps - can you suggest any?

shameless plug, [wf-recorder](https://github.com/ammen99/wf-recorder)

> Fair enough. :-) There are two darker grey strips within the overall lighter grey strip above the two left tiles on the bottom row, but it's a bit of a strain to see ....

I noticed these two stripes, but I believe it is just a bug in the application - it is part of the main surface of the application, so it must have drawn these two strips.

**Comment by flexibeast** [comment]
> shameless plug, [wf-recorder](https://github.com/ammen99/wf-recorder)

Excellent! Here's the result:

https://github.com/WayfireWM/wayfire/assets/190652/f53bed51-459c-401c-a182-7680d1edd4a3

(GitHub claims the video is corrupt, but locally mpv is able to play it - hope it works locally for you also.)

The sort of flickering i mean appears above the bottom-right tile in the right-hand grid.

**Comment by soreau** [comment]
> (GitHub claims the video is corrupt, but locally mpv is able to play it - hope it works locally for you also.)

This should be fixed with the default settings in latest wf-recorder git.

**Comment by soreau** [comment]
I can reproduce the problem with tetravex. It seems the client doesn't send damage requests for the region it changed, around the edges of the tiles. The colors at the latter part show the damage regions the client is giving for the surface. It uploads a new parent buffer with modified contents but never gets communicated to wayfire from the client, so the area isn't damaged.

https://github.com/WayfireWM/wayfire/assets/1450125/795ca2fb-c579-4ead-83c6-e3770c6a2225

**Comment by soreau** [comment]
This [tetravex patch](https://termbin.com/924e) fixes the problem here.

**Comment by ammen99** [comment]
For some context: we tried dumping the buffers we receive from tetravex, and they themselves turned out to contain visual artifacts, i.e the bug likely happened in the client. Probably tetravex picks a different theme when running under wayfire or something of the sort, in any case, likely not a Wayfire bug.

Nice find @soreau !

**Comment by flexibeast** [comment]
Thanks for looking into this! Unfortunately the above patch doesn't work for me; i'll check out gnome-tetravex's [bug tracker](https://gitlab.gnome.org/GNOME/gnome-tetravex/-/issues), and maybe open an issue, if there isn't one already.

**Comment by soreau** [comment]
@flexibeast I am quite confident the patch solves the issue, you might make sure the patch is applied and double check the binary built is the one being run.

**Comment by flexibeast** [comment]
@soreau Well, initially i tried using Gentoo's user patch system, but for some reason it didn't work with this particular ebuild - the patch wasn't being applied. So added `keepwork` to FEATURES, ran `ebuild [ebuild] configure`, removed those two lines, ran `ebuild [ebuild] merge`, checked the relevant file in the `src/` directory to confirm the lines hadn't been added back in, then ran `gnome-tetravex` - and there's only one such binary on my system.

**Comment by soreau** [comment]
Perhaps it's the case that you have to build latest gnome-tetravex from gitlab. At least that's where the patch should apply.

**Comment by flexibeast** [comment]
Sorry, i wasn't clear: when i say "wouldn't apply", i don't mean that applying the hunk *failed*, i mean applying the patch wasn't even attempted - normally one will get a message that an attempt is being made to apply the patch, regardless of whether it subsequently succeeds or fails. Hence me taking the "one problem at a time" approach, and changing the lines manually.

That said, it's true that i wasn't making the change in the source at HEAD - the latest ebuild is 3.38.2, and i see from the gnome-tetravex Issues page that there hasn't been a release in quite some time, so i'll try cloning the repo and building from HEAD with that change.

**Comment by flexibeast** [comment]
```
gnome-tetravex/src/ $ grep 'reset_clip' theme-extrusion.vala | wc -l
0
gnome-tetravex/src/ $ ls -l /usr/local/bin/gnome-tetravex 
-rwxr-xr-x 1 root root 1275728 Feb  6 20:11 /usr/local/bin/gnome-tetravex*
```

Running `/usr/local/bin/gnome-tetravex`, the artifacts are definitely still present.

**Comment by ammen99** [comment]
I just tested soreau's patch, the issues were fixed.

@flexibeast can you try capturing the contents of tetravex like we did? See the `tetravex-stuff` branch in this repo. Build it, set tetravex to autostart (and only it), run wayfire, launch tetravex and repro the bug. You should get a bunch of images in the current directory, which are snapshots of the contents of tetravex. If you notice any issues in them, then it is a client bug and you might need even more patches.

**Comment by flexibeast** [comment]
i'm certainly happy to try to do so, but sorry, i don't understand this:

> Build it, set tetravex to autostart (and only it), run wayfire, launch tetravex and repro the bug

Do you mean build Wayfire from the `textravex-stuff` branch? If i'm setting tetravex to autostart (presumably via `wayfire.ini`), why do you mention launching it manually?

**Comment by flexibeast** [comment]
This sequence demonstrates some of what i'm seeing, with various internal border changes between certain tiles from image to image. 

![img33](https://github.com/WayfireWM/wayfire/assets/190652/a465cf45-9578-4660-8abb-a0951c76ae54)
![img34](https://github.com/WayfireWM/wayfire/assets/190652/5c2511da-c938-4a4c-be29-90f1d7ece83f)
![img35](https://github.com/WayfireWM/wayfire/assets/190652/0d9e255c-c605-4ff4-ba17-97c55d33919d)
![img36](https://github.com/WayfireWM/wayfire/assets/190652/6e69b9f1-11ab-4a9e-88f3-aa0a7596b99c)
![img37](https://github.com/WayfireWM/wayfire/assets/190652/60dc3c27-885c-48fe-ad27-3503c36b9024)
![img38](https://github.com/WayfireWM/wayfire/assets/190652/0088b2e3-9457-42b6-a615-568acaa8ff3c)
![img39](https://github.com/WayfireWM/wayfire/assets/190652/78883803-623d-4fc1-8c3a-c3d17bc61c71)

**Comment by ammen99** [comment]
@flexibeast Those are the pngs generated by Wayfire, not screenshots, right? If that is the case, this is a bug in tetravex and will require a fix on the client side, because these borders are given to wayfire from the client itself.

**Comment by flexibeast** [comment]
Correct, those are the PNGs generated by Wayfire. Okay, thanks - i'll open a bug in the gnome-tetravex repo.

**Comment by soreau** [comment]
@flexibeast I think you might be using 'NeoRetro' or some theme other than 'Extrusion'. If you look at the patch, you can see it only patches `src/theme-extrusion.vala`. You should grep for reset_clip in all of the source directory. Anyway, I did all of this and reproduced the issue with 'NeoRetro' and fixed it.

TL;DR: [This new patch](https://termbin.com/q7fq) should fix the problem.

**Comment by flexibeast** [comment]
@soreau: Ah, i'd not noticed (or i'd forgotten) that it was possible to run gnome-tetravex with a different theme, so it didn't register that the original patch was simply patching a specific theme that i wasn't necessarily using.

Yep, removing `reset_clip` from `theme-neoretro.vala` appears to have fixed the issue for me. :-) Thanks to both you and @ammen99!

**Comment by soreau** [comment]
For what it's worth, the issue happens in weston too, further showing it's a client bug. Additionally, regarding cairo_reset_clip(), the [cairo docs](https://www.cairographics.org/manual/cairo-cairo-t.html#cairo-reset-clip) state:

"Note that code meant to be reusable should not call cairo_reset_clip() as it will cause results unexpected by higher-level code which calls cairo_clip(). Consider using cairo_save() and cairo_restore() around cairo_clip() as a more robust means of temporarily restricting the clip region."

**Comment by flexibeast** [comment]
@soreau: 👍 Thanks for taking the time to explain what's happening in more detail - i appreciate you and @ammen99 helping me learn stuff that will (hopefully!) help me produce better bug reports in the future. :-)
